cellCamera: Improvements and refactoring

elad335 opened the pull request:
* Fixed cellCameraEnd memleak.
* Fixed many error checks of null handler.
* Fixed cellCameraGetBufferSize. (override width/height, not bytesize)
* Fixed cellCameraIsAttached, remove hacks!
* Fixed cellCameraOpenEx. (check version!)
* Fixed cellCameraIsAvailable.
* Fixed data races.
* Added missing nullptr check in cellCameraGetBufferInfo.
* Simplify common error checking.
* Simplify read attribute.
* Simplify and atomic-ize cellCamera state management.

This pull request is also needed for savestates.

elad335 commented:
Wait with merging, I need some more cleanup in cellCamera.

sampletext32 commented on rpcs3/Emu/Cell/Modules/cellCamera.h:
```diff
@@ -429,12 +428,20 @@ class camera_context
 	{
 		u32 v1, v2;
+
+		using enbale_bitcopy = std::true_type;
```
Maybe you meant `enable`?)

elad335 commented:
* Fixed cellCameraEnd memleak.
* Fixed many error checks of null handler.
* Fixed cellCameraGetBufferSize. (override width/height, not bytesize)
* Fixed cellCameraIsAttached, remove hacks!
* Fixed cellCameraOpenEx. (check version!)
* Fixed cellCameraIsAvailable.
* Simplify common error checking.
* Simplify and atomic-ize cellCamera state management.

Megamouse commented on rpcs3/Emu/Cell/Modules/cellCamera.cpp:
```diff
@@ -769,33 +710,21 @@ error_code cellCameraGetBufferSize(s32 dev_num, vm::ptr<CellCameraInfoEx> info)
 		return CELL_CAMERA_ERROR_ALREADY_OPEN;
 	}
 
-	if (!info)
```
Changing the order of error checks is just wrong

elad335 replied:
Fixed.

Megamouse commented on rpcs3/Emu/Cell/Modules/cellCamera.cpp:
```diff
@@ -115,6 +116,40 @@ static bool check_dev_num(s32 dev_num)
+
+	if (func())
+	{
+		return CELL_CAMERA_ERROR_PARAM;
```
This assumes that the order never changes in any function, which is naive

elad335 replied:
It does not assume anything, its code. It just helps with common situation.

Megamouse commented on rpcs3/Emu/Cell/Modules/cellCamera.cpp:
```diff
@@ -115,6 +116,40 @@ static bool check_dev_num(s32 dev_num)
+
+	switch (state)
+	{
+	case camera_state::detached: return not_an_error(CELL_CAMERA_ERROR_DEVICE_NOT_FOUND);
```
Why is is this notanerror

elad335 replied:
Because it's a null handler error.

Megamouse commented on rpcs3/Emu/Cell/Modules/cellCamera.cpp:
```diff
@@ -280,14 +315,15 @@ error_code cellCameraInit()
-	if (g_camera.init)
+	const bool is_null = g_cfg.io.camera == camera_handler::null;
+
+	if (!g_camera.state.compare_and_swap_test(camera_state::uninit, is_null ? camera_state::detached : camera_state::closed))
```
Why is it initialized here before the function is done?

elad335 replied:
Why not?

elad335 replied:
Less LOC.

elad335 commented on rpcs3/Emu/Cell/Modules/cellCamera.cpp:
```diff
@@ -513,11 +546,16 @@ error_code cellCameraGetDeviceGUID(s32 dev_num, vm::ptr<u32> guid)
 		return CELL_CAMERA_ERROR_NOT_INIT;
 	}
 
+	if (!guid)
```
It does check nullptr.

Megamouse commented on rpcs3/Emu/Cell/Modules/cellCamera.cpp:
```diff
@@ -115,6 +116,40 @@ static bool check_dev_num(s32 dev_num)
+	}
+
+	if (func())
+	{
```
I'm not a fan of what you're doing here tbh.
1. It never reaches this point if the camera state is correct. So it is only checked if the camera state is wrong anyway.
2. It is kinda spaghetti code and hard to read in the flow of the functions
3. It adds another abstraction that you need to wrap your head around if you try to match assembly to code for example

So I personally would prefer if this function stayed a state check, and the param checks would remain conventional.

elad335 replied:
Oh, I see the ordering bug, will fix.
Although it packs common logic which in return reduces amount of LOC significantly. IMO it should be better documented as an optional ""optimization"" for code size. When this optimization cannot be followed one shall  return to the manual coding of checks.

Megamouse replied:
I appreciate the attempt to reduce LOC.
But i think in this case it's not really necessary, especially since it meddles with history as well

Megamouse replied:
And it makes this PR harder to review as well 💩

elad335 replied:
The main reason I'm modifying it in the first place is to remove unneeded variables for savestates (better to remove before PR is merged), history will be modified either way.
Everything in addition is a bonus.

elad335 replied:
Hmm maybe I can unwrap it relatively easily.

elad335 commented:
Fixed review comments.

Megamouse commented:
This is now very outdated

elad335 commented:
I don't even remember what it does, rip.
